### DISCUSSION TITLE
Static IRAM overrun compiling esp32 with IDF 5.2

### DISCUSSION THREAD

**Discussion by dannybr** [micropython/micropython#14263] [Core Development]
I think I'm doing everything by the book, but still failing: by 80 bytes :-)

/Users/danny/esp/.espressif5.2/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: micropython.elf section `.iram0.text' will not fit in region `iram0_0_seg'
/Users/danny/esp/.espressif5.2/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: IRAM0 segment data does not fit.
/Users/danny/esp/.espressif5.2/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: region `iram0_0_seg' overflowed by 80 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
HINT: The applications static IRAM usage is larger than the available IRAM size.
For more information on how to reduze IRAM usage run 'idf.py docs -sp api-guides/performance/ram-usage.html#optimizing-iram-usage' 
ninja failed with exit code 1, output of the command is in the /Users/danny/micropython/micropython/ports/esp32/build-ESP32_GENERIC/log/idf_py_stderr_output_28978 and /Users/danny/micropython/micropython/ports/esp32/build-ESP32_GENERIC/log/idf_py_stdout_output_28978
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [all] Error 1
danny@macmini esp32 % idf.py docs -sp api-guides/performance/ram-usage.html#optimizing-iram-usage
Executing action: docs
Opening documentation in the default browser:
https://docs.espressif.com/projects/esp-idf/en/v5.2/esp32/api-guides/performance/ram-usage.html#optimizing-iram-usage

**Comment by SquirtleSquadLeader** [comment]
I am having a similar issue trying to compile for UM_FEATHERS2.  

```
(esptoolenv) pi@pi400:~/Documents/micropython/ports/esp32 $ idf.py -p /dev/ttyACM0 flash
Executing action: flash
Running ninja in directory /home/pi/Documents/micropython/ports/esp32/build
Executing "ninja flash"...
[1/1] cd /home/pi/Documents/micropython/ports/esp32/build/b...nts/micropython/ports/esp32/build/bootloader/bootloader.bin
Bootloader binary size 0x5ac0 bytes. 0x1540 bytes (19%) free.
[5/12] cd /home/pi/Documents/micropython/ports/esp32/build/...ome/pi/Documents/micropython/ports/esp32/boards/manifest.py
[6/9] Linking CXX executable micropython.elf
FAILED: micropython.elf 

... omitting output

/home/pi/.espressif/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: micropython.elf section `.iram0.text' will not fit in region `iram0_0_seg'
/home/pi/.espressif/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: IRAM0 segment data does not fit.
/home/pi/.espressif/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/13.2.0/../../../../xtensa-esp-elf/bin/ld: region `iram0_0_seg' overflowed by 84 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
HINT: The applications static IRAM usage is larger than the available IRAM size.
For more information on how to reduze IRAM usage run 'idf.py docs -sp api-guides/performance/ram-usage.html#optimizing-iram-usage' 
ninja failed with exit code 1, output of the command is in the /home/pi/Documents/micropython/ports/esp32/build/log/idf_py_stderr_output_31065 and /home/pi/Documents/micropython/ports/esp32/build/log/idf_py_stdout_output_31065
```

**Comment by andrewleech** [comment]
I saw this recently myself when testing IDF 5.2 for esp32c6 support. This new version is clearly larger and doesn't fit the current micropython build settings for the original esp32.

I don't know if anyone is working on a fix for it yet, I don't have much time myself.

Note IDF 5.0.4 is the only officially supported version currently. Basic 5.2 support was recently contributed but the official builds / releases are still with 5.0.4 until compatibility issues such as this are ironed out!

**Comment by dannybr** [comment]
Removing the ethernet hardware support did it.

**Reply by ma261065** [reply]
How do you remove the ethernet hardware support?

**Comment by dannybr** [comment]
In ports/esp32, run idf.py menuconfig and look for 
components -> Ethernet and disable spi ethernet support
You can then run idf.py build

**Comment by GCTechnology** [comment]
I got the same problem... Checkout to `IDF 5.2` and run a 'make' command under `./ports/esp32`

**Comment by karunt** [comment]
Using ESP-IDF 5.2.0 and disabled spi ethernet support, but made no difference whatsoever. I'm still over by 584 bytes. I'm using ESP32-WROOM-32D module with 4MB flash. Any help would be appreciated. Here's my output log:

> [1/8] Generating ../../partition_table/partition-table.bin
Partition table binary generated. Contents:
*******************************************************************************
# ESP-IDF Partition Table
# Name, Type, SubType, Offset, Size, Flags
nvs,data,nvs,0xa000,20K,
phy_init,data,phy,0xf000,4K,
factory,app,factory,0x10000,3M,
storage,data,spiffs,0x310000,512K,
*******************************************************************************
[1/1] cmd.exe /C "cd /D C:\Users\ESP_code\bu...rs/ESP_code/build/bootloader/bootloader.bin" 
Bootloader binary size 0x6910 bytes. 0x6f0 bytes (6%) free.
[3/6] Linking CXX executable espcode.elf
FAILED: espcode.elf
cmd.exe /C "cd . && C:\Users\ESP\.espressif\tools\xtensa-esp32-elf\esp-12.2.0_20230208\xtensa-esp32-elf\bin\xtensa-esp32-elf-g++.exe -mlongcalls -Wno-frame-address -Wl,--cref -Wl,--defsym=IDF_TARGET_ESP32=0 -Wl,--Map=C:/Users/ESP_code/build/epuzz.map -Wl,--no-warn-rwx-segments -fno-rtti -fno-lto -Wl,--gc-sections -Wl,--warn-common -T esp32.peripherals.ld -T esp32.rom.ld -T esp32.rom.api.ld -T esp32.rom.libgcc.ld -T esp32.rom.newlib-data.ld -T esp32.rom.syscalls.ld -T esp32.rom.newlib-funcs.ld -T memory.ld -T sections.ld @CMakeFiles\espcode.elf.rsp -o espcode.elf  && cd ."
c:/users/esp/.espressif/tools/xtensa-esp32-elf/esp-12.2.0_20230208/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld.exe: espcode.elf section `.iram0.text' will not fit in region `iram0_0_seg'
c:/users/esp/.espressif/tools/xtensa-esp32-elf/esp-12.2.0_20230208/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld.exe: IRAM0 segment data does not fit.
c:/users/esp/.espressif/tools/xtensa-esp32-elf/esp-12.2.0_20230208/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld.exe: region `iram0_0_seg' overflowed by 584 bytes
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
HINT: The applications static IRAM usage is larger than the available IRAM size.

**Reply by karunt** [reply]
I reduced the amount of IRAM in use by enabling CONFIG_RINGBUF_PLACE_FUNCTIONS_INTO_FLASH and CONFIG_RINGBUF_PLACE_ISR_FUNCTIONS_INTO_FLASH. Works if you're not using the ringbuf component in your code. Not sure if that's the best way to resolve the issue, but I suppose I'll find out if/when something crashes.
